pkg/controller/workloadspread/workloadspread_controller.go

```diff
@@ -391,35 +397,66 @@ func (r *ReconcileWorkloadSpread) getSuitableSubsetForOldPod(ws *appsv1alpha1.Wo
 		return nil, err
 	}
 
+	var maxPreferredScore int64 = -1
+	var favoriteSubset *appsv1alpha1.WorkloadSpreadSubset
 	for i := range ws.Spec.Subsets {
 		subset := &ws.Spec.Subsets[i]
 		// in case of that this pod was scheduled to the node which matches a subset of workloadSpread
-		// TODO: check the patch of subset, one can have subset that only have patch but without any affinity or toleration changes.
-		if matched, err := matchesSubset(pod, node, subset); err != nil {
+		matched, preferredScore, err := matchesSubset(pod, node, subset)
+		if err != nil {
 			// requiredSelectorTerm field was validated at webhook stage, so this error should not occur
 			// this error should not be returned, because it is a non-transient error
 			klog.Errorf("unexpected error occurred when matching pod (%s/%s) with subset, please check requiredSelectorTerm field of subset (%s) in WorkloadSpread (%s/%s), err: %s",
 				pod.Namespace, pod.Name, subset.Name, ws.Namespace, ws.Name, err.Error())
-		} else if matched {
-			value, _ := json.Marshal(&wsutil.InjectWorkloadSpread{
-				Name:   ws.Name,
-				Subset: subset.Name,
-			})
-			patch, _ := json.Marshal(map[string]interface{}{
-				"metadata": map[string]interface{}{
-					"annotations": map[string]string{
-						wsutil.MatchedWorkloadSpreadSubsetAnnotations: string(value),
-					},
-				},
-			})
-			if err = r.Patch(context.TODO(), pod, client.RawPatch(types.StrategicMergePatchType, patch)); err != nil {
-				// no need to retry, this operation will be retried in the next reconcile
-				// this failure would not affect the subsequent logic
-				klog.Errorf(`Failed to patch "matched-workloadspread:{Name: %s, Subset: %s} annotation" for pod (%s/%s), but will retry in the next reconcile, err: %s`,
-					ws.Name, subset.Name, pod.Namespace, pod.Name, err.Error())
+		}
+		// select the most favorite subsets for the pod by matching subset.PreferredNodeSelectorTerms
+		if matched && preferredScore > maxPreferredScore {
+			favoriteSubset = subset
+			maxPreferredScore = preferredScore
+		}
+	}
+
+	// patch MatchedWorkloadSpreadSubsetAnnotations to the pod;
+	// select labels/annotations form favoriteSubset.patch, and patch them to the pod;
+	if favoriteSubset != nil {
+		var patchLabels map[string]string
+		patchAnnotations := make(map[string]string)
+		patchMetadata := make(map[string]interface{})
+
+		// decode favoriteSubset.patch.raw and add their labels/annotations to the patch
+		if favoriteSubset.Patch.Raw != nil {
+			patchField := &corev1.Pod{}
```

Comment:
I don't really agree with the patchField used here, for:

1. Now we get the `labels` and `annotations` from it, what if there are other fields like `finalizers`?
2. If users write some special semantics of strategic patch in it, such as `... {"foo": null}...` to remove a key-value in labels/annotations or `$deleteFromPrimitiveList/finalizers` to remove finalizers, they will be lost from the unmarshal.

IMO, we should unmarshal the raw into `map[string]interface{}`. Then you get/set the `metadata.annotations` from it or initialize the fields, and submit it as the patch body.

Reply:
1. Do you think that the `labels` and `annotations` **should not** be patched to pod?
2. Thanks. I will fix it.